<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Specificity Functions · siRNATools</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">siRNATools</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../instal_guide/">Installation Guide</a></li><li><span class="tocitem">Specificity</span><ul><li><a class="tocitem" href="../spec_guide/">Specificity Guide</a></li><li class="is-active"><a class="tocitem" href>Specificity Functions</a><ul class="internal"><li><a class="tocitem" href="#Reference-Sequence-1"><span>Reference Sequence</span></a></li><li><a class="tocitem" href="#Genome-Search-1"><span>Genome Search</span></a></li></ul></li></ul></li><li><span class="tocitem">Activity</span><ul><li><a class="tocitem" href="../act_guide/">Activity Guide</a></li><li><a class="tocitem" href="../act_index/">Activity Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Specificity</a></li><li class="is-active"><a href>Specificity Functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Specificity Functions</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/maxpwilson/siRNATools.jl/blob/master/docs/src/man/spec_index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Specificity-Functions-1"><a class="docs-heading-anchor" href="#Specificity-Functions-1">Specificity Functions</a><a class="docs-heading-anchor-permalink" href="#Specificity-Functions-1" title="Permalink"></a></h1><ul><li><a href="#siRNATools.Specificity.ReferenceSequence"><code>siRNATools.Specificity.ReferenceSequence</code></a></li><li><a href="#siRNATools.Specificity.Calculate_Specificity"><code>siRNATools.Specificity.Calculate_Specificity</code></a></li><li><a href="#siRNATools.Specificity.calculate_Peq"><code>siRNATools.Specificity.calculate_Peq</code></a></li><li><a href="#siRNATools.Specificity.compress_genome_matches"><code>siRNATools.Specificity.compress_genome_matches</code></a></li><li><a href="#siRNATools.Specificity.decode_refseq"><code>siRNATools.Specificity.decode_refseq</code></a></li><li><a href="#siRNATools.Specificity.decode_refseq_partial"><code>siRNATools.Specificity.decode_refseq_partial</code></a></li><li><a href="#siRNATools.Specificity.download_RefSeq"><code>siRNATools.Specificity.download_RefSeq</code></a></li><li><a href="#siRNATools.Specificity.encode_refseq"><code>siRNATools.Specificity.encode_refseq</code></a></li><li><a href="#siRNATools.Specificity.final_calc"><code>siRNATools.Specificity.final_calc</code></a></li><li><a href="#siRNATools.Specificity.find_genome_matches"><code>siRNATools.Specificity.find_genome_matches</code></a></li><li><a href="#siRNATools.Specificity.find_match_sequences"><code>siRNATools.Specificity.find_match_sequences</code></a></li><li><a href="#siRNATools.Specificity.get_refseq_pos"><code>siRNATools.Specificity.get_refseq_pos</code></a></li><li><a href="#siRNATools.Specificity.mismatch_positions"><code>siRNATools.Specificity.mismatch_positions</code></a></li><li><a href="#siRNATools.Specificity.motif_to_transcript_match"><code>siRNATools.Specificity.motif_to_transcript_match</code></a></li><li><a href="#siRNATools.Specificity.process_RefSeq"><code>siRNATools.Specificity.process_RefSeq</code></a></li><li><a href="#siRNATools.Specificity.reverse_complement"><code>siRNATools.Specificity.reverse_complement</code></a></li><li><a href="#siRNATools.Specificity.save_RefSeq"><code>siRNATools.Specificity.save_RefSeq</code></a></li></ul><h2 id="Reference-Sequence-1"><a class="docs-heading-anchor" href="#Reference-Sequence-1">Reference Sequence</a><a class="docs-heading-anchor-permalink" href="#Reference-Sequence-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.ReferenceSequence" href="#siRNATools.Specificity.ReferenceSequence"><code>siRNATools.Specificity.ReferenceSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Structure for reference sequences.  Compresses RNA data into 2 bits of information from the 8 of a normal character string.  Can only use bases A, C, G, U.</p><ul><li>A =&gt; 00</li><li>C =&gt; 01</li><li>G =&gt; 10</li><li>U =&gt; 11</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L5-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.get_refseq_pos" href="#siRNATools.Specificity.get_refseq_pos"><code>siRNATools.Specificity.get_refseq_pos</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">get_refseq_pos(::ReferenceSequence, ::Int)</code></pre><p>Function to retrieve bit value of base at specificied position in a reference sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L55-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.encode_refseq" href="#siRNATools.Specificity.encode_refseq"><code>siRNATools.Specificity.encode_refseq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">encode_refseq(::String) :: ReferenceSequence</code></pre><p>Function that takes a string containing only A, C, G, and U and encodes it into a ReferenceSequence type</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L67-L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.decode_refseq" href="#siRNATools.Specificity.decode_refseq"><code>siRNATools.Specificity.decode_refseq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">decode_refseq(::ReferenceSequence) :: String</code></pre><p>Function takes a ReferenceSequence type and returns a String containing the RNA bases.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L94-L98">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.decode_refseq_partial" href="#siRNATools.Specificity.decode_refseq_partial"><code>siRNATools.Specificity.decode_refseq_partial</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">decode_refseq_partial(::ReferenceSequence, ::UnitRange) :: String</code></pre><p>Function which takes a ReferenceSequence type and outputs specificied range of bases as a String.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L116-L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.download_RefSeq" href="#siRNATools.Specificity.download_RefSeq"><code>siRNATools.Specificity.download_RefSeq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">download_RefSeq(::UnitRange{Int64}=1:8, ::String=PATH)</code></pre><p>Downloads mRNA reference sequence from ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/ to the PATH folder.  Defaults to downloading 8 files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L129-L133">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.process_RefSeq" href="#siRNATools.Specificity.process_RefSeq"><code>siRNATools.Specificity.process_RefSeq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">process_RefSeq(::UnitRange{Int64}=1:8, ::String=PATH)</code></pre><p>Processes raw gzipped fasta files into DataFrame and saves it as a CSV in the PATH folder</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L173-L177">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.save_RefSeq" href="#siRNATools.Specificity.save_RefSeq"><code>siRNATools.Specificity.save_RefSeq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">save_RefSeq(::String=PATH)</code></pre><p>Saves relevant data structures from the processes mRNA reference sequence for use in searches.</p><ul><li>TranscriptGene =&gt; dictionary of Transcripts to Genes</li><li>GeneTranscripts =&gt; dictionary of Genes to Transcripts</li><li>allT =&gt; dictionary of Transcript name to base sequence as String</li><li>allRefSeq =&gt; dictionary of Transcript name to base sequence as ReferenceSequence</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L193-L201">source</a></section></article><h2 id="Genome-Search-1"><a class="docs-heading-anchor" href="#Genome-Search-1">Genome Search</a><a class="docs-heading-anchor-permalink" href="#Genome-Search-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.Calculate_Specificity" href="#siRNATools.Specificity.Calculate_Specificity"><code>siRNATools.Specificity.Calculate_Specificity</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">Calculate_Specificity(patterns, excluded_gene=&quot;&quot;, rg=2:18, verbose=true) :: DataFrame</code></pre><p>Function takes as input the pattern to be searched against the genome, the excluded gene to be ignored, the range of pattern being used, and a boolean which controls whether progress bars will be shown.  Output is a DataFrame with a column for the pattern, the number of genes with minimum mismatch distance of 0-4 and the specificity score.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L435-L441">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.reverse_complement" href="#siRNATools.Specificity.reverse_complement"><code>siRNATools.Specificity.reverse_complement</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">reverse_complement(::String) :: String</code></pre><p>Takes the reverse complement of an RNA string</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L224-L228">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.calculate_Peq" href="#siRNATools.Specificity.calculate_Peq"><code>siRNATools.Specificity.calculate_Peq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">calculate_Peq(::String) :: Array{UInt64, 1}</code></pre><p>Takes as input an RNA strand and translates it into an array of Unsigned Integers each one representing one letter and each bit referring to a position.   Used as a pre-processing step for Myers searching algorithm.  Maximum length of searched pattern is 64.</p><p><strong>Example</strong></p><p>AACGCU becomes [110000, 001010, 000100, 000001]</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L242-L251">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.motif_to_transcript_match" href="#siRNATools.Specificity.motif_to_transcript_match"><code>siRNATools.Specificity.motif_to_transcript_match</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">motif_to_transcript_match(Peq::Array{Unit64, 1}, m::Int64, refseq::ReferenceSequence, min_K::Int) :: Array{Tuple{UInt64, UInt64}}</code></pre><p>Takes as input Peq calculated in <a href="man/@ref">calculate_Peq</a>, m = the length of the pattern to search for, the ReferenceSequence to search in, and min_k =  the maximum distance to output.  Implements the string matching algorithm described in <em>A Fast Bit-Vector Algorithm for Approximate String Matching Based on Dynamic Programming</em> by Gene Myers.  Output is a tuple of all Levenshtein distances less than min_k and their positions in refseq.   </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L266-L272">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.find_match_sequences" href="#siRNATools.Specificity.find_match_sequences"><code>siRNATools.Specificity.find_match_sequences</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_match_sequences(motif::String, sequence::String, mismatches::Int) :: Array{String, 1}</code></pre><p>Function takes as input a motif, sequence, and number of mismatches to search for.  Output is an array of all substrings of sequence which have a  Hamming distance of exactly mismatches to motif. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L305-L310">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.mismatch_positions" href="#siRNATools.Specificity.mismatch_positions"><code>siRNATools.Specificity.mismatch_positions</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">mismatch_positions(::String, ::String) :: Array{Int, 1}</code></pre><p>Function returns an array of all positions in which strings used as input differ.  Strings must be the same length.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L319-L323">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.find_genome_matches" href="#siRNATools.Specificity.find_genome_matches"><code>siRNATools.Specificity.find_genome_matches</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">find_genome_matches(pattern::String, excluded_gene::String = &quot;&quot;, verbose::Bool = true, minimum_matches = 5) :: Array{Tuple{String, Int64}}</code></pre><p>Function takes as input a pattern to search the genome for, excluded<em>gene to exclude the gene of interest from the search, verbose set to true displays a progress bar showing progress of search, and minimum\</em>matches which is the amount of mismatches searched for + 1.  Output is an array of tuples of transcript names and the lowest Hamming distance found to the pattern within that transcript.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L337-L343">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.compress_genome_matches" href="#siRNATools.Specificity.compress_genome_matches"><code>siRNATools.Specificity.compress_genome_matches</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">compress_genome_matches(::Array{Tuple{String, Int64}}) :: Dict{String, Array{Int64, 1}}</code></pre><p>Function takes as input the output of <a href="man/@ref">find_genome_matches</a> and collapses it into a dictionary of Gene names and an array of all lowest Hamming mismatch  of the transcripts of that gene.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L368-L373">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="siRNATools.Specificity.final_calc" href="#siRNATools.Specificity.final_calc"><code>siRNATools.Specificity.final_calc</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">final_calc(pattern::String, raw_data::Array{Tuple{String, Int64}}, compressed_data::Dict{String, Array{Int64, 1}})</code></pre><p>Function takes as input the pattern being searched for, the raw_data from <a href="man/@ref">find_genome_matches</a>, and the compressed data from <a href="man/@ref">compress_genome_matches</a>. Ouput is a Tuple containing the mismatch_counts dictionary which adds up the number of genes with a minimum Hamming distance of 0-4, and the specificity score.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maxpwilson/siRNATools.jl/blob/b0074234822b14324764e35131b5f65d2bcebab3/src/Specificity.jl#L388-L393">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../spec_guide/">« Specificity Guide</a><a class="docs-footer-nextpage" href="../act_guide/">Activity Guide »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 3 January 2020 14:37">Friday 3 January 2020</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
